design/kiota-e2e.md

```diff
@@ -0,0 +1,215 @@
+-   `@microsoft/msgraph-sdk-javascript-beta` ([npm](https://www.npmjs.com/package/@microsoft/msgraph-sdk-javascript-type)) - The service library for making fluent calls to Microsoft Graph beta.
+
+You can use [npm](https://www.npmjs.com) to install the Microsoft Graph Javascript Service Library :
+
+```Shell
+npm install @microsoft/msgraph-sdk-javascript --save
+npm install @microsoft/msgraph-sdk-javascript-beta --save-dev
+npm install @microsoft/kiota-authentication-azure --save
+```
+
+### Importing the right functionalities from the Graph Javascript Service Library
+
+```typescript
+// App.ts
+
+import { Client, User, Message, BodyType } from "@microsoft/msgraph-sdk-javascript";
+import { AzureIdentityAuthenticationProvider } from "@microsoft/kiota-authentication-azure";
+import { DeviceCodeCredential } from "@azure/identity";
+```
+
+### Setting up the app to work with the Graph Javascript Service Library
+
+```typescript
+const deviceCodeCredentials = new DeviceCodeCredential({
+	tenantId: "b61f9af1-d6cf-4cc0-a6f6-befb38bc00ed",
+	clientId: "bde251a6-0ef9-42a8-a40b-9ad9bb594b2c",
+});
+
+const scopes = ["User.Read", "Mail.Send"];
+
+const graphClient = Client.init({
+	authenticationTokenProvider: new AzureIdentityAuthenticationProvider(deviceCodeCredentials, scopes),
+});
+```
+
+### Calling Microsoft Graph via the Graph Javascript Service Library
+
+The core value of the Microsoft Graph Javascript Service Library is the availability of the Fluent API. This provides developers with typechecking and discoverability when building on Microsoft Graph. This Fluent API is made available via the Request Builder concept and highlights the full spectrum of capabilities on Microsoft Graph.
+
+Models are also available in this package and should reflect the underlying version of Graph we are targeting. Developers using our types should do it in a way that will be familiar to how they are used to with other SDKs and APIs. We should provide an easy-to-use model that feels natural in the Javascript world. This means we should not be forcing developers to use Classes and / or other structures that will make their code less natural to them.
+
+We have an open issue about the above topic that can be followed here : [TypeScript - Use of interface models instead of class models](https://github.com/microsoft/kiota/issues/1013)
+
+```typescript
+const me = await getMe();
+const meRaw = await getMeRaw();
+await sendMail();
+await sendMailRaw();
+
+// The types returned by the fluent API should be the same as the .api() area. It should also be the same (or at least very similar) as the current @microsoft/microsoft-graph-types to offer seamless migration.
+async function getMe(): Promise<User | undefined> {
+	return await graphClient.me.get();
+}
+
+// Allowing raw calls (using the .api() method instead of the full fluent API) is important for migration purposes and cases we don't know the resource beforehands (thinking Graph Explorer, mgt-get, etc.)
```

Comment:
I understand the migration scenario, but could you please elaborate the case of not knowing the resource upfront?

Reply:
what do you mean by not knowing the resource upfront?

Reply:
Oh, so you mean cases where the request is built on runtime rather than design-time, correct?

Reply:
Yes, in this example, the response type would be any/unknown. But I believe our developers will want to have a typings experience even using raw http calls. Similar to how you can leverage the SDK today.

Reply:
If you use a raw HTTP call but know the URL upfront, then by extension you'll know the shape of data returned by the API and can cast the response to the correct type, correct? I think the challenge is the odd case where you construct the URL on runtime and you don't know what data you'd be getting, but I don't think that's a common usecase.

Reply:
I understand your point. Do you think we should literally avoid the types here? It feels there is value for developers with a preference with using endpoint vs. the fluent API...

Reply:
Do you mean: "what if folks wanted to achieve what they're having today on v3, with v4 but without loading the fluent API, and having just the core functionality and typings"?

Reply:
Exactly!

Reply:
If we want to support this scenario, wouldn't that mean that we need to offer two SDKs (core and service) after all which brings us back to the original argument?

Reply:
It's not something we'd publicize that much, but I would say yes. This needs to be supported for these advanced scenarios. I'm happy to park this idea and listen for feedback though... Which would allow us to focus on a single path (the service library) assuming it delivers all the underlying core features (which it should).